File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/CS/Messages/Msg.cs
namespace Db4objects.Db4o.CS.Messages
{
	/// <summary>Messages for Client/Server Communication</summary>
	public class Msg : Sharpen.Lang.ICloneable
	{
		internal static int _idGenerator = 1;

		private static Db4objects.Db4o.CS.Messages.Msg[] _messages = new Db4objects.Db4o.CS.Messages.Msg
			[60];

		internal int _msgID;

		internal string _name;

		internal Db4objects.Db4o.Transaction _trans;

		public static readonly Db4objects.Db4o.CS.Messages.MsgD CLASS_NAME_FOR_ID = new Db4objects.Db4o.CS.Messages.MClassNameForID
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg CLOSE = new Db4objects.Db4o.CS.Messages.Msg
			("CLOSE");

		public static readonly Db4objects.Db4o.CS.Messages.Msg COMMIT = new Db4objects.Db4o.CS.Messages.MCommit
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg COMMIT_OK = new Db4objects.Db4o.CS.Messages.MCommitOK
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD CREATE_CLASS = new Db4objects.Db4o.CS.Messages.MCreateClass
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg CURRENT_VERSION = new Db4objects.Db4o.CS.Messages.Msg
			("VERSION");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD DELETE = new Db4objects.Db4o.CS.Messages.MDelete
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg ERROR = new Db4objects.Db4o.CS.Messages.Msg
			("ERROR");

		public static readonly Db4objects.Db4o.CS.Messages.Msg FAILED = new Db4objects.Db4o.CS.Messages.Msg
			("FAILED");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD GET_ALL = new Db4objects.Db4o.CS.Messages.MGetAll
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD GET_CLASSES = new Db4objects.Db4o.CS.Messages.MGetClasses
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD GET_INTERNAL_IDS = new Db4objects.Db4o.CS.Messages.MGetInternalIDs
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg GET_THREAD_ID = new Db4objects.Db4o.CS.Messages.Msg
			("GET_THREAD_ID");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD ID_LIST = new Db4objects.Db4o.CS.Messages.MsgD
			("ID_LIST");

		public static readonly Db4objects.Db4o.CS.Messages.Msg IDENTITY = new Db4objects.Db4o.CS.Messages.Msg
			("IDENTITY");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD LENGTH = new Db4objects.Db4o.CS.Messages.MsgD
			("LENGTH");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD LOGIN = new Db4objects.Db4o.CS.Messages.MsgD
			("LOGIN");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD LOGIN_OK = new Db4objects.Db4o.CS.Messages.MsgD
			("LOGIN_OK");

		public static readonly Db4objects.Db4o.CS.Messages.Msg NULL = new Db4objects.Db4o.CS.Messages.Msg
			("NULL");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECT_BY_UUID = new Db4objects.Db4o.CS.Messages.MObjectByUuid
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgObject OBJECT_TO_CLIENT = new 
			Db4objects.Db4o.CS.Messages.MsgObject();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_FETCH = new Db4objects.Db4o.CS.Messages.MObjectSetFetch
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_FINALIZED = new 
			Db4objects.Db4o.CS.Messages.MsgD("OBJECTSET_FINALIZED");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_GET_ID = new Db4objects.Db4o.CS.Messages.MObjectSetGetId
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_INDEXOF = new Db4objects.Db4o.CS.Messages.MObjectSetIndexOf
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_RESET = new Db4objects.Db4o.CS.Messages.MObjectSetReset
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD OBJECTSET_SIZE = new Db4objects.Db4o.CS.Messages.MObjectSetSize
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg OK = new Db4objects.Db4o.CS.Messages.Msg
			("OK");

		public static readonly Db4objects.Db4o.CS.Messages.Msg PING = new Db4objects.Db4o.CS.Messages.Msg
			("PING");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD PREFETCH_IDS = new Db4objects.Db4o.CS.Messages.MPrefetchIDs
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgObject QUERY_EXECUTE = new 
			Db4objects.Db4o.CS.Messages.MQueryExecute();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD QUERY_RESULT = new Db4objects.Db4o.CS.Messages.MsgD
			("QUERY_RESULT");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD RAISE_VERSION = new Db4objects.Db4o.CS.Messages.MsgD
			("RAISE_VERSION");

		public static readonly Db4objects.Db4o.CS.Messages.MsgBlob READ_BLOB = new Db4objects.Db4o.CS.Messages.MReadBlob
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD READ_BYTES = new Db4objects.Db4o.CS.Messages.MReadBytes
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD READ_MULTIPLE_OBJECTS = new 
			Db4objects.Db4o.CS.Messages.MReadMultipleObjects();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD READ_OBJECT = new Db4objects.Db4o.CS.Messages.MReadObject
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD RELEASE_SEMAPHORE = new Db4objects.Db4o.CS.Messages.MReleaseSemaphore
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg ROLLBACK = new Db4objects.Db4o.CS.Messages.MRollback
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD SET_SEMAPHORE = new Db4objects.Db4o.CS.Messages.MSetSemaphore
			();

		public static readonly Db4objects.Db4o.CS.Messages.Msg SUCCESS = new Db4objects.Db4o.CS.Messages.Msg
			("SUCCESS");

		public static readonly Db4objects.Db4o.CS.Messages.MsgD SWITCH_TO_FILE = new Db4objects.Db4o.CS.Messages.MsgD
			("SWITCH_F");

		public static readonly Db4objects.Db4o.CS.Messages.Msg SWITCH_TO_MAIN_FILE = new 
			Db4objects.Db4o.CS.Messages.Msg("SWITCH_M");

		public static readonly Db4objects.Db4o.CS.Messages.Msg TA_BEGIN_END_SET = new Db4objects.Db4o.CS.Messages.MTaBeginEndSet
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD TA_DELETE = new Db4objects.Db4o.CS.Messages.MTaDelete
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD TA_DONT_DELETE = new Db4objects.Db4o.CS.Messages.MTaDontDelete
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD TA_IS_DELETED = new Db4objects.Db4o.CS.Messages.MTaIsDeleted
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD USER_MESSAGE = new Db4objects.Db4o.CS.Messages.MUserMessage
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD USE_TRANSACTION = new Db4objects.Db4o.CS.Messages.MUseTransaction
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgBlob WRITE_BLOB = new Db4objects.Db4o.CS.Messages.MWriteBlob
			();

		public static readonly Db4objects.Db4o.CS.Messages.MWriteNew WRITE_NEW = new Db4objects.Db4o.CS.Messages.MWriteNew
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgObject WRITE_UPDATE = new Db4objects.Db4o.CS.Messages.MWriteUpdate
			();

		public static readonly Db4objects.Db4o.CS.Messages.MsgD WRITE_UPDATE_DELETE_MEMBERS
			 = new Db4objects.Db4o.CS.Messages.MWriteUpdateDeleteMembers();

		internal Msg()
		{
			_msgID = _idGenerator++;
			_messages[_msgID] = this;
		}

		internal Msg(string aName) : this()
		{
			_name = aName;
		}

		public Db4objects.Db4o.CS.Messages.Msg Clone(Db4objects.Db4o.Transaction a_trans)
		{
			Db4objects.Db4o.CS.Messages.Msg msg = null;
			try
			{
				msg = (Db4objects.Db4o.CS.Messages.Msg)MemberwiseClone();
				msg._trans = a_trans;
			}
			catch (Sharpen.Lang.CloneNotSupportedException)
			{
			}
			return msg;
		}

		public sealed override bool Equals(object obj)
		{
			if (this == obj)
			{
				return true;
			}
			if (obj == null || obj.GetType() != this.GetType())
			{
				return false;
			}
			return _msgID == ((Db4objects.Db4o.CS.Messages.Msg)obj)._msgID;
		}

		internal virtual void FakePayLoad(Db4objects.Db4o.Transaction a_trans)
		{
			_trans = a_trans;
		}

		/// <summary>
		/// dummy method to allow clean override handling
		/// without casting
		/// </summary>
		public virtual Db4objects.Db4o.YapReader GetByteLoad()
		{
			return null;
		}

		internal string GetName()
		{
			if (_name == null)
			{
				return GetType().FullName;
			}
			return _name;
		}

		protected virtual Db4objects.Db4o.Transaction Transaction()
		{
			return _trans;
		}

		protected virtual Db4objects.Db4o.YapFile File()
		{
			return (Db4objects.Db4o.YapFile)Stream();
		}

		protected virtual Db4objects.Db4o.YapStream Stream()
		{
			return Transaction().Stream();
		}

		protected virtual object StreamLock()
		{
			return Stream().Lock();
		}

		protected virtual Db4objects.Db4o.Config4Impl Config()
		{
			return Stream().Config();
		}

		/// <summary>server side execution</summary>
		/// <param name="serverThread">TODO</param>
		public virtual bool ProcessAtServer(Db4objects.Db4o.CS.YapServerThread serverThread
			)
		{
			return false;
		}

		public static Db4objects.Db4o.CS.Messages.Msg ReadMessage(Db4objects.Db4o.Transaction
			 a_trans, Db4objects.Db4o.Foundation.Network.IYapSocket sock)
		{
			Db4objects.Db4o.YapWriter reader = new Db4objects.Db4o.YapWriter(a_trans, Db4objects.Db4o.YapConst
				.MESSAGE_LENGTH);
			if (!reader.Read(sock))
			{
				return null;
			}
			Db4objects.Db4o.CS.Messages.Msg message = _messages[reader.ReadInt()].ReadPayLoad
				(a_trans, sock, reader);
			return message;
		}

		internal virtual Db4objects.Db4o.CS.Messages.Msg ReadPayLoad(Db4objects.Db4o.Transaction
			 a_trans, Db4objects.Db4o.Foundation.Network.IYapSocket sock, Db4objects.Db4o.YapReader
			 reader)
		{
			a_trans = CheckParentTransaction(a_trans, reader);
			return Clone(a_trans);
		}

		protected virtual Db4objects.Db4o.Transaction CheckParentTransaction(Db4objects.Db4o.Transaction
			 a_trans, Db4objects.Db4o.YapReader reader)
		{
			if (reader.ReadByte() == Db4objects.Db4o.YapConst.SYSTEM_TRANS && a_trans.ParentTransaction
				() != null)
			{
				return a_trans.ParentTransaction();
			}
			return a_trans;
		}

		internal void SetTransaction(Db4objects.Db4o.Transaction aTrans)
		{
			_trans = aTrans;
		}

		public sealed override string ToString()
		{
			return GetName();
		}

		public void Write(Db4objects.Db4o.YapStream stream, Db4objects.Db4o.Foundation.Network.IYapSocket
			 sock)
		{
			lock (sock)
			{
				try
				{
					sock.Write(PayLoad()._buffer);
					sock.Flush();
				}
				catch
				{
				}
			}
		}

		public virtual Db4objects.Db4o.YapWriter PayLoad()
		{
			Db4objects.Db4o.YapWriter writer = new Db4objects.Db4o.YapWriter(Transaction(), Db4objects.Db4o.YapConst
				.MESSAGE_LENGTH);
			writer.WriteInt(_msgID);
			return writer;
		}
	}
}

File: db4oj/core/src/com/db4o/internal/cs/messages/Msg.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.internal.cs.messages;

import java.io.*;

import com.db4o.*;
import com.db4o.ext.*;
import com.db4o.foundation.*;
import com.db4o.foundation.network.*;
import com.db4o.internal.*;
import com.db4o.internal.cs.*;

/**
 * Messages for Client/Server Communication
 */
public abstract class Msg implements Cloneable {

	static int _messageIdGenerator = 1;
	private static Msg[] _messages = new Msg[60];

	int _msgID;
	String _name;
	private Transaction _trans;
	private MessageDispatcher _messageDispatcher;
	

	public static final MClassNameForID CLASS_NAME_FOR_ID = new MClassNameForID();
	public static final MClose CLOSE = new MClose();
    public static final MCommit COMMIT = new MCommit();
    public static final MCommittedCallBackRegistry COMMITTED_CALLBACK_REGISTER = new MCommittedCallBackRegistry();
    public static final MCommittedInfo COMMITTED_INFO = new MCommittedInfo();
    public static final MCommitSystemTransaction COMMIT_SYSTEMTRANS = new MCommitSystemTransaction();
	public static final MCreateClass CREATE_CLASS = new MCreateClass();
	public static final MClassMeta CLASS_META = new MClassMeta();
	public static final MVersion CURRENT_VERSION = new MVersion();
	public static final MDb4oException DB4O_EXCEPTION = new MDb4oException();
	public static final MDelete DELETE = new MDelete();
	public static final MError ERROR = new MError();
	public static final MFailed FAILED = new MFailed();
	public static final MGetAll GET_ALL = new MGetAll();
	public static final MGetClasses GET_CLASSES = new MGetClasses();
	public static final MGetInternalIDs GET_INTERNAL_IDS = new MGetInternalIDs();
	public static final MGetThreadID GET_THREAD_ID = new MGetThreadID();
	public static final MIDList ID_LIST = new MIDList();
	public static final MIdentity IDENTITY = new MIdentity();
	public static final MLength LENGTH = new MLength();
    public static final MLogin LOGIN = new MLogin();
    public static final MLoginOK LOGIN_OK = new MLoginOK();
	public static final MNull NULL = new MNull();
	public static final MObjectByUuid OBJECT_BY_UUID = new MObjectByUuid();
	public static final MsgObject OBJECT_TO_CLIENT = new MsgObject();
	public static final MObjectSetFetch OBJECTSET_FETCH = new MObjectSetFetch();
	public static final MObjectSetFinalized OBJECTSET_FINALIZED = new MObjectSetFinalized();
	public static final MObjectSetGetId OBJECTSET_GET_ID = new MObjectSetGetId();
	public static final MObjectSetIndexOf OBJECTSET_INDEXOF = new MObjectSetIndexOf();
	public static final MObjectSetReset OBJECTSET_RESET = new MObjectSetReset();
	public static final MObjectSetSize OBJECTSET_SIZE = new MObjectSetSize();
	public static final MOK OK = new MOK();
	public static final MPing PING = new MPing();
	public static final MPong PONG = new MPong();
	public static final MPrefetchIDs PREFETCH_IDS = new MPrefetchIDs();
	public static final MProcessDeletes PROCESS_DELETES = new MProcessDeletes();
	public static final MQueryExecute QUERY_EXECUTE = new MQueryExecute();
	public static final MQueryResult QUERY_RESULT = new MQueryResult();
	public static final MRaiseVersion RAISE_VERSION = new MRaiseVersion();
	public static final MReadBlob READ_BLOB = new MReadBlob();
	public static final MReadBytes READ_BYTES = new MReadBytes();
	public static final MReadMultipleObjects READ_MULTIPLE_OBJECTS = new MReadMultipleObjects();
	public static final MReadObject READ_OBJECT = new MReadObject();
	public static final MReleaseSemaphore RELEASE_SEMAPHORE = new MReleaseSemaphore();
	public static final MRollback ROLLBACK = new MRollback();
	public static final MSetSemaphore SET_SEMAPHORE = new MSetSemaphore();
	public static final MSuccess SUCCESS = new MSuccess();
	public static final MSwitchToFile SWITCH_TO_FILE = new MSwitchToFile();
	public static final MSwitchToMainFile SWITCH_TO_MAIN_FILE = new MSwitchToMainFile();
	public static final MTaDelete TA_DELETE = new MTaDelete();
	public static final MTaIsDeleted TA_IS_DELETED = new MTaIsDeleted();
	public static final MUserMessage USER_MESSAGE = new MUserMessage();
	public static final MUseTransaction USE_TRANSACTION = new MUseTransaction();
	public static final MWriteBlob WRITE_BLOB = new MWriteBlob();
	public static final MWriteNew WRITE_NEW = new MWriteNew();
	public static final MWriteUpdate WRITE_UPDATE = new MWriteUpdate();
	public static final MWriteUpdateDeleteMembers WRITE_UPDATE_DELETE_MEMBERS = new MWriteUpdateDeleteMembers();
	public static final MWriteBatchedMessages WRITE_BATCHED_MESSAGES = new MWriteBatchedMessages();

	Msg() {
		_msgID = _messageIdGenerator++;
		_messages[_msgID] = this;
	}

	Msg(String aName) {
		this();
		_name = aName;
	}
	
	public static Msg getMessage(int id) {
		return _messages[id];
	}
	
	public final Msg publicClone() {
		try {
			return (Msg)clone();
		} catch (CloneNotSupportedException e) {
			Exceptions4.shouldNeverHappen();
			return null;
		}
	}
	
	public final boolean equals(Object obj) {
		if(this==obj) {
			return true;
		}
		if(obj==null||obj.getClass()!=this.getClass()) {
			return false;
		}
	    return _msgID == ((Msg) obj)._msgID;
	}
	
	public int hashCode() {
		return _msgID;
	}

	void fakePayLoad(Transaction a_trans) {
	    _trans = a_trans;
		// do nothing
	}

	/**
	 * dummy method to allow clean override handling
	 * without casting
	 */
	public Buffer getByteLoad() {
		return null;
	}

	final String getName() {
		if (_name == null) {
			return getClass().getName();
		}
		return _name;
	}
	
	protected LocalTransaction serverTransaction(){
		return (LocalTransaction)_trans;
	}

	protected Transaction transaction() {
		return _trans;
	}
	
	protected LocalObjectContainer file(){
		return (LocalObjectContainer)stream();
	}
	
	protected ObjectContainerBase stream(){
	    return transaction().stream();
	}
	
	protected Object streamLock(){
		return stream().lock();
	}
	
	protected Config4Impl config(){
		return stream().config();
	}
	
    protected static StatefulBuffer readMessageBuffer(Transaction trans, Socket4 sock) throws IOException {
		return readMessageBuffer(trans, sock, Const4.MESSAGE_LENGTH);
    }

	protected static StatefulBuffer readMessageBuffer(Transaction trans, Socket4 sock, int length) throws IOException {
		StatefulBuffer buffer = new StatefulBuffer(trans, length);		
        int offset = 0;
        while (length > 0) {
            int read = sock.read(buffer._buffer, offset, length);
			if(read<0) {
				return null;
			}
            offset += read;
            length -= read;
        }
		return buffer;
	}


	public static final Msg readMessage(MessageDispatcher messageDispatcher, Transaction trans, Socket4 sock) throws IOException {
		StatefulBuffer reader = readMessageBuffer(trans, sock);
		if (null == reader) {
			return null;
		}
		Msg message = _messages[reader.readInt()].readPayLoad(messageDispatcher, trans, sock, reader);
		if (Debug.messages) {
			System.out.println(message + " arrived at " + trans.stream());
		}
		return message;
	}

	Msg readPayLoad(MessageDispatcher messageDispatcher, Transaction a_trans, Socket4 sock, Buffer reader)
		throws IOException {
		Msg msg = publicClone();
		msg.setMessageDispatcher(messageDispatcher);
		msg.setTransaction(checkParentTransaction(a_trans, reader));
	    return msg;
	}

	protected final Transaction checkParentTransaction(Transaction a_trans, Buffer reader) {
		if(reader.readByte() == Const4.SYSTEM_TRANS && a_trans.parentTransaction() != null){
	        return a_trans.parentTransaction();
	    }
		return a_trans;
	}

	public final void setTransaction(Transaction aTrans) {
		_trans = aTrans;
	}

	final public String toString() {
		return getName();
	}
	

	public void write(Msg msg) {
		_messageDispatcher.write(msg);
	}
	
	public void writeException(Db4oException e) {
		write(DB4O_EXCEPTION.getWriterForSingleObject(transaction(), e));
	}
	
	public void respondInt(int response){
    	write(ID_LIST.getWriterForInt(transaction(), response));
    }
	
	public final void write(ObjectContainerBase stream, Socket4 sock) {
		if (null == stream) {
			throw new ArgumentNullException();
		}
		if (null == sock) {
			throw new ArgumentNullException();
		}
		if (Debug.fakeServer) {
		    ObjectContainerBase i_stream = null;
			if (stream == DebugCS.serverStream) {
				i_stream = DebugCS.clientStream;
			} else {
				i_stream = DebugCS.serverStream;
			}
			setTransaction(i_stream.getTransaction());
			fakePayLoad(i_stream.getTransaction());
			if (stream == DebugCS.serverStream) {
                DebugCS.clientMessageQueue.add(this);
			} else {
				((ServerSideMessage)this).processAtServer();
			}
		} else {
			synchronized (sock) {
				try {
					if (Debug.messages) {
						System.out.println(this +" sent by " + stream);
					}
					sock.write(payLoad()._buffer);
					sock.flush();
				} catch (Exception e) {
                    
                    // Note: It is not sufficient to catch IoException only
                    // here. .NET will throw different exceptions if a socket
                    // is disconnected.
                    
					// TODO: handle more softly in YapClient, maybe
					// try to reconnect
                    
				}
			}
		}
	}

	public StatefulBuffer payLoad() {
		StatefulBuffer writer = new StatefulBuffer(transaction(), Const4.MESSAGE_LENGTH);
		writer.writeInt(_msgID);
		return writer;
	}

	
	public MessageDispatcher messageDispatcher() {
		return _messageDispatcher;
	}
	
	public ServerMessageDispatcher serverMessageDispatcher() {
		if(_messageDispatcher instanceof ServerMessageDispatcher) {
			return (ServerMessageDispatcher) _messageDispatcher;	
		}
		throw new IllegalStateException();
	}

	public ClientMessageDispatcher clientMessageDispatcher() {
		if(_messageDispatcher instanceof ClientMessageDispatcher) {
			return (ClientMessageDispatcher) _messageDispatcher;	
		}
		throw new IllegalStateException();
	}
	
	public void setMessageDispatcher(MessageDispatcher messageDispatcher) {
		_messageDispatcher = messageDispatcher;
	}
	
	public void logMsg(int msgCode, String msg) {
		stream().logMsg(msgCode, msg);
	}

}